doc/educational_decoder/zstd_decompress.c

```diff
@@ -1303,145 +1272,148 @@ static void execute_sequences(frame_context_t *const ctx, ostream_t *const out,
 
     for (size_t i = 0; i < num_sequences; i++) {
         const sequence_command_t seq = sequences[i];
-
         {
-            // If the sequence asks for more literals than are left, the
-            // sequence must be corrupted
-            if (seq.literal_length > IO_istream_len(&litstream)) {
-                CORRUPTION();
-            }
+            const u32 literals_size = copy_literals(seq, &litstream, out);
+            total_output += literals_size;
+        }
+        size_t offset = compute_offset(seq, offset_hist);
+
+        size_t match_length = seq.match_length;
 
-            u8 *const write_ptr = IO_write_bytes(out, seq.literal_length);
-            const u8 *const read_ptr =
-                    IO_read_bytes(&litstream, seq.literal_length);
-            // Copy literals to output
-            memcpy(write_ptr, read_ptr, seq.literal_length);
+        execute_match_copy(ctx, offset, match_length, total_output, out);
 
-            total_output += seq.literal_length;
-        }
+        total_output += match_length;
+    }
 
-        size_t offset;
-
-        // Offsets are special, we need to handle the repeat offsets
-        if (seq.offset <= 3) {
-            // "The first 3 values define a repeated offset and we will call
-            // them Repeated_Offset1, Repeated_Offset2, and Repeated_Offset3.
-            // They are sorted in recency order, with Repeated_Offset1 meaning
-            // 'most recent one'".
-
-            // Use 0 indexing for the array
-            u32 idx = seq.offset - 1;
-            if (seq.literal_length == 0) {
-                // "There is an exception though, when current sequence's
-                // literals length is 0. In this case, repeated offsets are
-                // shifted by one, so Repeated_Offset1 becomes Repeated_Offset2,
-                // Repeated_Offset2 becomes Repeated_Offset3, and
-                // Repeated_Offset3 becomes Repeated_Offset1 - 1_byte."
-                idx++;
-            }
+    // Copy any leftover literals
+    {
+        size_t len = IO_istream_len(&litstream);
```

Comment:
minor, but let's detail since this is a learning opportunity : 

This part is slightly more complex than necessary.
I would expect here to find something like `copy_literals()`.

`copy_literals()` cannot be used "as is" because it requires a `sequence_command_t` input type, and you only have a `size_t len`.
However, `copy_literals()` effectively only needs `seq.literal_length`.
It could be refactored to only use this length, instead of the full `sequence_command_t`.
In which case, it becomes usable for leftover literals.

If you believe it's valuable to not get into the details of `sequence_command_t`, you could nest it under some `copy_literals_from_seq()`, which just extracts `seq.literal_length` and pass it to the new `copy_literals()`.
However, I see just above: `size_t match_length = seq.match_length;`
which means the structure of `sequence_command_t` is already a dependency of `execute_sequences()`. So it doesn't seem worthwhile to avoid it.